> 오류 처리는 중요하다. 하지만 로직을 헷갈리게 만드는 오류 처리는 나쁘다    
(Error handling is important, but if it obscures logic, it's wrong.)

### 리턴 코드 대신 Exceptions 를 사용하라
- 예전 프로그래밍 언어들은 Exceptions 를 제공하지 않았다.
- 그 경우, 개발자들은 에러 flag 를 set 하거나 에러 코드를 리턴, 호출하는 측에서 예외처리를 해줘야 했다.
- 하지만, 이러한 방식은 예외처리를 잊어버리기 쉽고, 로직을 헷갈리게 하기 쉽다.
- 그러므로 Exceptions 를 사용하자. 겉보기에만 아름다운 코드가 되는게 아니라 '실제 로직' 과 '예외처리' 부분이 나뉘어져 필요한 부분에 집중할 수 있게 된다.

```` java
// Bad
public class DeviceController {
  ...
  public void sendShutDown() {
    DeviceHandle handle = getHandle(DEV1);
    // Check the state of the device
    if (handle != DeviceHandle.INVALID) {
      // Save the device status to the record field
      retrieveDeviceRecord(handle);
      // If not suspended, shut down
      if (record.getStatus() != DEVICE_SUSPENDED) {
        pauseDevice(handle);
        clearDeviceWorkQueue(handle);
        closeDevice(handle);
      } else {
        logger.log("Device suspended. Unable to shut down");
      }
    } else {
      logger.log("Invalid handle for: " + DEV1.toString());
    }
  }
  ...
}
````

```` java
// Good
public class DeviceController {
  ...
  public void sendShutDown() {
    try {
      tryToShutDown();
    } catch (DeviceShutDownError e) {
      logger.log(e);
    }
  }
    
  private void tryToShutDown() throws DeviceShutDownError {
    DeviceHandle handle = getHandle(DEV1);
    DeviceRecord record = retrieveDeviceRecord(handle);
    pauseDevice(handle); 
    clearDeviceWorkQueue(handle); 
    closeDevice(handle);
  }
  
  private DeviceHandle getHandle(DeviceID id) {
    ...
    throw new DeviceShutDownError("Invalid handle for: " + id.toString());
    ...
  }
  ...
}
````

### Try-Catch-Finally 문을 먼저 써라
- try 문은 transaction 처럼 동작하는 실행 코드로, catch 문은 try 문에 관계없이 프로그램을 일관적인 상태로 유지하도록 한다.
- 이렇게 함으로써 코드의 'Scope 정의' 가 가능해진다.
- 예시 : 잘못된 input 을 넣을 경우, StorageException 을 제대로 던지는지 확인하는 테스트 코드를 작성해보자.

```` java
// Step 1: StorageException 을 던지지 않으므로 이 테스트는 실패한다.
@Test(expected = StorageException.class)
public void retrieveSectionShouldThrowOnInvalidFileName() {
    sectionStore.retrieveSection("invalid - file");
}
// throws 로 exception 을 호출한 곳으로 던지지 않음
public List<RecordedGrip> retrieveSection(String sectionName) {
    // dummy return until we have a real implementation
    return new ArrayList<RecordedGrip>();
}
````

```` java
// Step 2: 이제 테스트는 통과한다.
public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
      FileInputStream stream = new FileInputStream(sectionName)
    } catch (Exception e) {
      // StorageException 를 호출한 곳으로 던짐
      throw new StorageException("retrieval error", e);
    }
    return new ArrayList<RecordedGrip>();
}
````

```` java
// Step 3: Exception의 범위를 FileNotFoundException으로 줄여 정확히 어떤 Exception이 발생한지 체크하자.
public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
        FileInputStream stream = new FileInputStream(sectionName);
        stream.close();
    } catch (FileNotFoundException e) { // Step 2. 와 달리 exception 의 범위를 FileNotFoundException 으로 줄임
        throw new StorageException("retrieval error", e);
    }
    return new ArrayList<RecordedGrip>();
}
````

### Unchecked Exceptions 를 사용하라
- Checked Exception vs Unchecked Exception 
- 예외 처리에 드는 비용 대비 이득을 생각해봐야 한다.

```` plain
1. 특정 메소드에서 checked exception 을 throws 하고,
2. 3단계 (메소드 콜) 위의 메소드에서 그 exception 을 catch 한다면,
3. 모든 중간 단계 메소드에 exception 을 정의해야 한다. 
````

- 상위 레벨 메소드에서 하위 레벨 메소드의 디테일에 대해 알아야 하기 때문에 캡슐화 또한 깨진다 (Open Closed Principle - 개방 폐쇄 원칙)
- 필요한 경우, checked exception 을 사용해야 하지만, 일반적인 경우 득보다는 실이 많다.

### Exceptions 로 문맥을 제공하라
- 예외가 발생한 이유와 좀 더 구체적인 Exception 타입을 제공하라.

### 사용에 맞게 Exception 클래스를 선언하라
- Exception class 를 만드는 데 가장 중요한 것은 '어떤 방식으로 예외를 잡을까' 이다.
- Third Party 라이브러리를 사용하는 경우, 그것들을 Wrapping 함으로써 아래와 같은 장점이 생기게 된다.

```` plain
1. 라이브러리 교체 등의 변경이 있는 경우, 대응하기 쉬워진다.
2. 라이브러리를 쓰는 곳을 테스트할 경우, 해당 라이브러리를 가짜로 만들 수 있음으로써 테스트하기가 쉬워진다.
3. 라이브러이의 api 디자인에 종속적이지 않고, 내 입맛에 맞는 디자인을 적용할 수 있다.
````

- 보통 특정 부분의 코드에는 exception 하나로 충분히 예외처리를 할 수 있다.
- 하나의 exception 만 잡고, 나머지 하나는 다시 throw 하는 경우 등 정말 필요한 경우에만 다른 exception 클래스를 만들어 사용하자.   

- 예시 : 외부 API 클래스인 ACMEPort 클래스를 사용하는 상황을 살펴보자.

```` java
  // Bad
  // catch문의 내용이 거의 같다.
  ACMEPort port = new ACMEPort(12);
  try {
    port.open();
  } catch (DeviceResponseException e) {
    reportPortError(e);
    logger.log("Device response exception", e);
  } catch (ATM1212UnlockedException e) {
    reportPortError(e);
    logger.log("Unlock exception", e);
  } catch (GMXError e) {
    reportPortError(e);
    logger.log("Device response exception");
  } finally {
    ...
  }
````

```` java
  // Good
  // ACME 클래스를 LocalPort 클래스로 래핑해 new ACMEPort().open() 메소드에서 던질 수 있는 exception 들을 간략화
  
  LocalPort port = new LocalPort(12);
  try {
    port.open();
  } catch (PortDeviceFailure e) {
    reportError(e);
    logger.log(e.getMessage(), e);
  } finally {
    ...
  }
  
  -----------------------------------------
  
  public class LocalPort {
    private ACMEPort innerPort;
    public LocalPort(int portNumber) {
      innerPort = new ACMEPort(portNumber);
    }
    
    // ACMEPort 클래스를 LocalPort 로 Wrapping 한 후 ACMEPort 클래스에서 발생할 수 있는 예외들을 PortDeviceFailure 커스텀 예외 클래스로 던짐
    // 다양한 원인의 예외들을 자세한 메시지를 통해 예외로 던짐 (throw new PortDeviceFaulure(e);)
    public void open() {
      try {
        innerPort.open();
      } catch (DeviceResponseException e) {
        throw new PortDeviceFailure(e);
      } catch (ATM1212UnlockedException e) {
        throw new PortDeviceFailure(e);
      } catch (GMXError e) {
        throw new PortDeviceFailure(e);
      }
    }
    ...
  }
````

### 정상적인 상황을 정의하라 (Default 값을 설정하라)
- 일반적으로는 위에서 봤던 방식들이 유용하지만, catch 문에서 예외적인 상황(special case)을 처리해야 하는 경우, 코드가 더러워지는 일이 발생할 수 있다.
- 이런 경우, 마틴 파울러의 'Special Case Pattern' 을 사용하자.
    - 코드를 부르는 입장에서 예외적인 상황을 신경쓰지 않아도 된다.
    - 예외 상황은 special case object 내에 캡슐화 된다.
    
````java
  // Bad
  try {
    MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
    m_total += expenses.getTotal();
  } catch(MealExpensesNotFound e) {
    m_total += getMealPerDiem();
  }
```` 

````java
  // Good
  // caller logic.
  ...
  MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
  m_total += expenses.getTotal();
  ...
  
  // MealExpensesNotFound Exception 이 발생할 경우, 별도의 커스텀 예외 클래스인 PerDiemMealExpenses 를 통해서 getTotal 로직을 수행하도록 작성
  // 이로 인하여 'ExpenseReportDAO' 를 호출하는 쪽에서는 별도의 exception 처리를 하지 않아도 되므로 코드가 더 간결해질 수 있다.
  public class PerDiemMealExpenses implements MealExpenses {
    public int getTotal() {
      // return the per diem default
      // 
    }
  }
  
  // 이해를 돕기 위해 직접 추가한 클래스
  public class ExpenseReportDAO {
    ...
    public MealExpenses getMeals(int employeeId) {
      MealExpenses expenses;
      try {
        expenses = expenseReportDAO.getMeals(employee.getID());
      } catch(MealExpensesNotFound e) {
        expenses = new PerDiemMealExpenses();
      }
      
      return expenses;
    }
    ...
  }
````

### Null 을 리턴하지 마라
- Null 을 리턴하고 싶은 생각이 들면, 위의 Special Case Object 를 리턴하라
- Third Party 라이브러리에서 Null 을 리턴할 가능성이 있는 메서드가 있다면, Exception 을 던지거나 Special Case Object 를 리턴하는 메서드로 Wrapping 하라.

````java
  // BAD!!!!

  public void registerItem(Item item) {
    if (item != null) {
      // peristentStore 가 null 인 경우에 대한 예외처리가 없음
      // 만약 여기서 NullPointerException 이 발생했다면, 수십 단계 위의 메소드에서 처리해줘야 하나?
      ItemRegistry registry = peristentStore.getItemRegistry();
      if (registry != null) {
        Item existing = registry.getItem(item.getID());
        if (existing.getBillingPeriod().hasRetailOwner()) {
          existing.register(item);
        }
      }
    }
  }
  
  // 이 메소드의 문제점은 null 체크가 부족한게 아니라 null 체크가 너무 많다는 것이다.
````

````java
  // Bad
  List<Employee> employees = getEmployees();
  if (employees != null) {
    for(Employee e : employees) {
      totalPay += e.getPay();
    }
  }
  // getEmployees() 가 null 인 경우, null 을 리턴하게 된다.
````

````java
  // Good
  List<Employee> employees = getEmployees();
  for(Employee e : employees) {
    totalPay += e.getPay();
  }
  // getEmployees() 에서 employee 가 없어서 null 일 경우, Collections.emptyList() 를 통해 빈 리스트를 리턴한다.
  public List<Employee> getEmployees() {
    if( .. there are no employees .. ) {
      return Collections.emptyList();
    }
  }
````

### Null 을 넘기지 마라
- null 을 리턴하는 것도 나쁘지만, null 을 메서드로 넘기는 것은 더 나쁘다.
- null 을 메서드의 파라미터로 넣어야 하는 API 를 사용하는 경우가 아니면, null 을 메서드로 넘기지 마라.
- 일반적으로 대다수의 프로그래밍 언어들은 파라미터로 들어온 Null 에 대해 적절한 방법을 제공하지 못한다.
- 가장 이상적인 해법은 null 을 파라미터로 받지 못하게 하는 것이다.

```` java
// Bad
// calculator.xProjection(null, new Point(12, 13));
// 위와 같이 부를 경우 NullPointerException 발생
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    return (p2.x – p1.x) * 1.5;
  }
  ...
}

// Bad
// NullPointerException 은 안 나지만 윗 단계에서 InvalidArgumentException 이 발생할 경우 처리해줘야 함.
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    if(p1 == null || p2 == null){
      throw InvalidArgumentException("Invalid argument for MetricsCalculator.xProjection");
    }
    return (p2.x – p1.x) * 1.5;
  }
}

// Bad
// 좋은 명세이지만 첫번째 예시와 같이 NullPointerException 문제를 해결하지 못한다.
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    // 근본적인 원인인 NullPointerException 를 해결할 수 있지 못함
    // JDK 1.4 부터 지원하는 assert 키워드는 거짓일 경우, AssertionError 예외 발생
    assert p1 != null : "p1 should not be null";
    assert p2 != null : "p2 should not be null";
    
    return (p2.x – p1.x) * 1.5;
  }
}
````

### 결론
> 깨끗한 코드와 견고한 코드는 대립되는 목표가 아니다.   
예외 처리를 로직에서 제거하면, 각각에 대해 독립적인 사고가 가능해진다!